Client/Frontend/Browser/Authenticator.swift

```diff
@@ -48,6 +73,25 @@ class Authenticator {
         return self.promptForUsernamePassword(viewController, credentials: nil, protectionSpace: challenge.protectionSpace, loginsHelper: nil)
     }
 
+    private static func findValidLoginAmongstLogins(logins: Cursor<LoginData>, forProtectionSpace protectionSpace: NSURLProtectionSpace) -> LoginData? {
+        // A valid login is defined as one that matches the challenge protection space and is not malformed
+        return logins.filter { login in
+            return (login?.protectionSpace.`protocol` == protectionSpace.`protocol`) && !(login?.hasMalformedHostname ?? false)
+        } .flatMap { $0 } .first
+    }
+
+    private static func removeMalformedLoginsFromLogins(logins: Cursor<LoginData>, usingHelper loginsHelper: LoginsHelper) -> Success {
+        let malformedGUIDs: [GUID] = logins.map { login in
```

Comment:
`flatMap` *is* `map` + `flatten`, so you can rewrite this more simply.

```
logins.filter { $0.hasMalformedHostname }.map { $0.guid }
```

or

```
logins.flatMap { login in
    if login.hasMalformedHostname {
        return login.guid
    }
    return nil
}
```